Move all CacheStorageConnection callbacks to NativePromise

youennf opened the pull request:
#### d244968bca644b6ae5f293054c1c1ba757cef5a2
<pre>
Move all CacheStorageConnection callbacks to NativePromise
<a href="https://bugs.webkit.org/show_bug.cgi?id=274726">https://bugs.webkit.org/show_bug.cgi?id=274726</a>
<a href="https://rdar.apple.com/128755510">rdar://128755510</a>

Reviewed by Jean-Yves Avenard and Sihui Liu.

We move from completion handlers to native promise as a simplification.
We introduce a new enqueueTaskWhenSettled version that can take completion handler and a finalizer.
Covered by existing tests.

* Source/WebCore/Modules/cache/CacheStorageConnection.h:
(WebCore::CacheStorageConnection::clearMemoryRepresentation):
(WebCore::CacheStorageConnection::engineRepresentation):
* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::queryCache):
(WebCore::DOMCache::batchDeleteOperation):
(WebCore::DOMCache::batchPutOperation):
* Source/WebCore/Modules/cache/DOMCacheEngine.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::DOMCacheStorage::retrieveCaches):
(WebCore::DOMCacheStorage::doOpen):
(WebCore::DOMCacheStorage::doRemove):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::retrieveCaches):
(WebCore::WorkerCacheStorageConnection::retrieveRecords):
(WebCore::WorkerCacheStorageConnection::batchDeleteOperation):
(WebCore::WorkerCacheStorageConnection::batchPutOperation):
(WebCore::WorkerCacheStorageConnection::retrieveCachesCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::retrieveRecordsCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::deleteRecordsCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::putRecordsCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::clearPendingRequests): Deleted.
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::enqueueTaskWhenSettled):
* Source/WebCore/page/CacheStorageProvider.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::clearCacheStorageMemoryRepresentation):
(WebCore::Internals::cacheStorageEngineRepresentation):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::prepareForDestruction):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::retrieveCaches):
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchDeleteOperation):
(WebKit::WebCacheStorageConnection::batchPutOperation):
(WebKit::WebCacheStorageConnection::clearMemoryRepresentation):
(WebKit::WebCacheStorageConnection::engineRepresentation):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:

Canonical link: <a href="https://commits.webkit.org/279505@main">https://commits.webkit.org/279505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d048531b8ca2c911101521226311540ba29928ce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53580 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6086 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56860 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4306 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55886 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4109 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43435 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2827 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55678 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31223 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46309 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24570 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28054 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3631 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2462 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58457 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28739 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3856 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50840 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29947 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46481 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50181 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11696 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30877 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/dom/ScriptExecutionContext.h:
```diff
@@ -409,7 +409,7 @@ class ScriptExecutionContext : public SecurityContext, public TimerAlignment {
 
     void checkConsistency() const;
-    RefCountedSerialFunctionDispatcher& nativePromiseDispatcher();
+    WEBCORE_EXPORT RefCountedSerialFunctionDispatcher& nativePromiseDispatcher();
```
Is this needed?

youennf replied:
Yes, for using enqueueTaskWhenSettled within Internals.

szewai commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6247,36 +6247,40 @@ void Internals::sendH2Ping(String url, DOMPromiseDeferred<IDLDouble>&& promise)
 
     if (!m_cacheStorageConnection) {
-        if (auto* page = contextDocument()->page())
+        if (RefPtr page = contextDocument()->page())
```
`document->page()`

youennf replied:
ok

szewai commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6247,36 +6247,40 @@ void Internals::sendH2Ping(String url, DOMPromiseDeferred<IDLDouble>&& promise)
+        if (RefPtr page = contextDocument()->page())
             m_cacheStorageConnection = page->cacheStorageProvider().createCacheStorageConnection();
         if (!m_cacheStorageConnection)
             return;
```
Should we reject the promise on the early return?

youennf replied:
We could but I do not think this is needed.

szewai commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6247,36 +6247,40 @@ void Internals::sendH2Ping(String url, DOMPromiseDeferred<IDLDouble>&& promise)
+        if (RefPtr page = contextDocument()->page())
             m_cacheStorageConnection = page->cacheStorageProvider().createCacheStorageConnection();
         if (!m_cacheStorageConnection)
             return;
```
Should we reject the promise on the early return?

youennf replied:
Ditto

szewai commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6247,36 +6247,40 @@ void Internals::sendH2Ping(String url, DOMPromiseDeferred<IDLDouble>&& promise)
-    m_cacheStorageConnection->engineRepresentation([promise = WTFMove(promise)](const String& result) mutable {
-        promise.resolve(result);
+    document->enqueueTaskWhenSettled(m_cacheStorageConnection->engineRepresentation(), TaskSource::DOMManipulation, [promise = WTFMove(promise)](auto&& result) mutable {
+        if (!result.has_value()) {
```
`if (!result)`

szewai commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
```diff
@@ -67,24 +67,32 @@ auto WebCacheStorageConnection::remove(WebCore::DOMCacheIdentifier cacheIdentifi
 {
-    connection().sendWithAsyncReply(Messages::NetworkStorageManager::CacheStorageAllCaches(origin, updateCounter), WTFMove(callback));
+    return connection().sendWithPromisedReply(Messages::NetworkStorageManager::CacheStorageAllCaches(origin, updateCounter))->whenSettled(RunLoop::current(), [](auto&& result) {
+        return result ? RetrieveCachesPromise::createAndSettle(WTFMove(result.value())) : RetrieveCachesPromise::createAndReject(WebCore::DOMCacheEngine::Error::Internal);
```
Why are we rejecting with `DOMCacheEngine::Error::Internal` instead of `result.error()`?

jyavenard replied:
> Why are we rejecting with `DOMCacheEngine::Error::Internal` instead of `result.error()`?

sendWithPromiseReply returns an Expected<Expected<T,F>, IPC::ErrorCode>

szewai reviewed (commented):
Test failure seems relevant

jyavenard commented on Source/WebCore/Modules/cache/CacheStorageConnection.h:
```diff
@@ -59,8 +63,10 @@ class CacheStorageConnection : public ThreadSafeRefCounted<CacheStorageConnectio
     // Used only for testing purposes.
-    virtual void clearMemoryRepresentation(const ClientOrigin&, DOMCacheEngine::CompletionCallback&& callback) { callback(DOMCacheEngine::Error::NotImplemented); }
-    virtual void engineRepresentation(CompletionHandler<void(const String&)>&& callback) { callback(String { }); }
+    using CompletionPromise = NativePromise<void, DOMCacheEngine::Error>;
```
I would use a `GenericPromise` here instead, considering the error code is ignored in the end, and we only care if it resolved or rejected.

jyavenard reviewed (approved):
LGTM with Sihui's comments applied.

I personally prefer
`if (result)` vs `if (result.has_value())` or

`return *result;` vs `return result.value()`

but seeing this is what the existing code was using, I guess it's fine like this.

szewai commented on Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
```diff
@@ -239,9 +241,9 @@ void DOMCacheStorage::doOpen(const String& name, DOMPromiseDeferred<IDLInterface
     }
 
-    context->enqueueTaskWhenSettled(m_connection->open(*origin(), name), TaskSource::DOMManipulation, [this, name, promise = WTFMove(promise), pendingActivity = makePendingActivity(*this), connectionStorageLock = makeUnique<ConnectionStorageLock>(m_connection.copyRef(), *origin())](const DOMCacheEngine::CacheIdentifierOrError& result) mutable {
+    context->enqueueTaskWhenSettled(m_connection->open(*origin(), name), TaskSource::DOMManipulation, [this, name, promise = WTFMove(promise), pendingActivity = makePendingActivity(*this), connectionStorageLock = makeUnique<ConnectionStorageLock>(m_connection.copyRef(), *origin())] (auto&& result) mutable {
```
This one does not have finalizer, is it expected?

youennf replied:
Yes, it is only necessary when capturing a completion handler.
Or if we decide to reject the promise when document goes away (I don't think we have a coherent approach in that area, whether to let the promise dangling or whether to reject).

szewai commented on Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
```diff
@@ -277,7 +279,7 @@ void DOMCacheStorage::doRemove(const String& name, DOMPromiseDeferred<IDLBoolean
     }
 
     scriptExecutionContext()->enqueueTaskWhenSettled(m_connection->remove(m_caches[position]->identifier()), TaskSource::DOMManipulation, [this, promise = WTFMove(promise), pendingActivity = makePendingActivity(*this)](const auto& result) mutable {
```
This one does not have finalizer, is it expected?

youennf replied:
Ditto

youennf committed:
Move all CacheStorageConnection callbacks to NativePromise
https://bugs.webkit.org/show_bug.cgi?id=274726
rdar://128755510

Reviewed by Jean-Yves Avenard and Sihui Liu.

We move from completion handlers to native promise as a simplification.
We introduce a new enqueueTaskWhenSettled version that can take completion handler and a finalizer.
Covered by existing tests.

* Source/WebCore/Modules/cache/CacheStorageConnection.h:
(WebCore::CacheStorageConnection::clearMemoryRepresentation):
(WebCore::CacheStorageConnection::engineRepresentation):
* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::queryCache):
(WebCore::DOMCache::batchDeleteOperation):
(WebCore::DOMCache::batchPutOperation):
* Source/WebCore/Modules/cache/DOMCacheEngine.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::DOMCacheStorage::retrieveCaches):
(WebCore::DOMCacheStorage::doOpen):
(WebCore::DOMCacheStorage::doRemove):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::retrieveCaches):
(WebCore::WorkerCacheStorageConnection::retrieveRecords):
(WebCore::WorkerCacheStorageConnection::batchDeleteOperation):
(WebCore::WorkerCacheStorageConnection::batchPutOperation):
(WebCore::WorkerCacheStorageConnection::retrieveCachesCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::retrieveRecordsCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::deleteRecordsCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::putRecordsCompleted): Deleted.
(WebCore::WorkerCacheStorageConnection::clearPendingRequests): Deleted.
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::enqueueTaskWhenSettled):
* Source/WebCore/page/CacheStorageProvider.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::clearCacheStorageMemoryRepresentation):
(WebCore::Internals::cacheStorageEngineRepresentation):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::prepareForDestruction):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::retrieveCaches):
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchDeleteOperation):
(WebKit::WebCacheStorageConnection::batchPutOperation):
(WebKit::WebCacheStorageConnection::clearMemoryRepresentation):
(WebKit::WebCacheStorageConnection::engineRepresentation):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:

Canonical link: https://commits.webkit.org/279505@main
